### DISCUSSION TITLE
getting an explenation rather than a function name in function_call_formatter

### DISCUSSION THREAD

**Discussion by dcssi** [zhudotexe/kani#18] [Q&A] [answered]
When defining a function, an explanation of its functionality is written in English  for the model to use to determine when the function should be called.  
When the model decides that it needs to call the function, it calls function_call_formatter with the message.  
The message contains just the name of the function and the parameters.

I think it would be beneficial, if the class FunctionCall also contains the English description the function has. 
The function name is good for python, but for the user we should use English.
I guess, I could implement this in my code by doing a lookup by function name.

**Comment by zhudotexe** [comment]
Hmm, could you post a code/pseudocode snippet that shows what the desired behaviour looks like? I'm not sure I understand the question.

**Reply by dcssi** [reply]
For example, lets look at  the 2_function_calling_weather.py example.   You got the function:
 get_weather
with the description
"""Get the current weather in a given location."""

So when you run it, in the history you have something like this:
User: is it cloudy in sf ?
Ai: Thinking (get_weather)
Ai: The weather in san fran...

function_call_formatter only has access to the function name and the parameters that are going to be passed to the function.  
I could setup a dictionary with function names as keys and a template to generate a message in the formatter.

I think that if we had access to the description in the formatter we can generate a better message than using the function name.  Also there is a lot more flexibility in the description compared to python syntax rules.

**Reply by zhudotexe** [reply, marked as answer]
Oh, I think I see. You can access a mapping of function names to `AIFunction` objects through the `Kani.functions` attribute.

For example, if you wanted to print the docstring of a function when the AI requested to call it, you could do:

```py
engine = ...
ai = MyKani(...)

def function_call_formatter(msg):
  func = ai.functions.get(message.function_call.name)
  if func is None:
    # raise some error - the AI requested a nonexistent function
  return f"The AI is calling {func.name}, which has this documentation:\n{func.desc}"

# ...
async for msg in ai.full_round_str(..., function_call_formatter=function_call_formatter):
  ...
```

**Reply by zhudotexe** [reply]
Alternatively, you could override [`Kani.do_function_call`](https://kani.readthedocs.io/en/latest/customization.html#handle-a-function-call).
